astropy/modeling/core.py

```diff
@@ -1763,6 +1763,148 @@ def rename(self, name):
+        input_units_allow_dimensionless : bool or dict, optional
+            Allow dimensionless input. If this is True, input values to evaluate will
+            gain the units specified in input_units. If this is a dictionary then it
+            should map input name to a bool to allow dimensionless numbers for that
+            input.
+
+        Returns
+        -------
+        `CompoundModel`
+            A `CompoundModel` composed of the current model plus
+            `~astropy.modeling.mappings.UnitsMapping` model(s) that attach the units.
+
+        Raises
+        ------
+        ValueError
+            If the current model already has units.
+
+        Examples
+        --------
+
+        Wrapping a unitless model to require and convert units:
+
+        >>> from astropy.modeling.models import Polynomial1D
+        >>> from astropy import units as u
+        >>> poly = Polynomial1D(1, c0=1, c1=2)
+        >>> model = poly.coerce_units((u.m,), (u.s,))
+        >>> model(u.Quantity(10, u.m))  # doctest: +FLOAT_CMP
+        <Quantity 21. s>
+        >>> model(u.Quantity(1000, u.cm))  # doctest: +FLOAT_CMP
+        <Quantity 21. s>
+        >>> model(u.Quantity(10, u.cm))  # doctest: +FLOAT_CMP
+        <Quantity 1.2 s>
+
+        Wrapping a unitless model but still permitting unitless input:
+
+        >>> from astropy.modeling.models import Polynomial1D
+        >>> from astropy import units as u
+        >>> poly = Polynomial1D(1, c0=1, c1=2)
+        >>> model = poly.coerce_units((u.m,), (u.s,), input_units_allow_dimensionless=True)
+        >>> model(u.Quantity(10, u.m))  # doctest: +FLOAT_CMP
+        <Quantity 21. s>
+        >>> model(10)  # doctest: +FLOAT_CMP
+        <Quantity 21. s>
+        """
+        from .mappings import UnitsMapping
+
+        result = self
+
+        if input_units is not None:
+            if self.input_units is not None:
+                model_units = self.input_units
+            else:
+                model_units = {}
+
+            for unit in [model_units.get(i) for i in self.inputs]:
+                if unit is not None and unit != dimensionless_unscaled:
+                    raise ValueError("Cannot specify input_units for model with existing input units")
+
+            if isinstance(input_units, dict):
+                if input_units.keys() != set(self.inputs):
+                    message = (
+                        f"""input_units keys ({", ".join(input_units.keys())}) """
+                        f"""do not match model inputs ({", ".join(self.inputs)})"""
+                    )
+                    raise ValueError(message)
+                input_units = [input_units[i] for i in self.inputs]
+
+            if len(input_units) != self.n_inputs:
+                message = (
+                    "input_units length does not match n_inputs: "
+                    f"expected {self.n_inputs}, received {len(input_units)}"
+                )
+                raise ValueError(message)
+
```

Comment:
Where is `return_units` handled? Was it lost during the rebase?

As it is the `mapping` ends up being `((unit, None),...)`  and this fails to serialize, e.g. the example in the docstring.
Could you add also a test to the tags directory that actually writes to an asdf file.

```

Reply:
Uh oh, I didn't think about that.  We can't currently serialize `None` values because asdf doesn't write nulls; instead it just drops them from the tree (see https://github.com/spacetelescope/asdf/issues/536).  I think I can make this work in the meantime by making `None` the default, I'll give that a try.

Reply:
`return_units` is handled on line 1874.

The `assert_roundtrip_tree` helper that I'm using in tags/transform/tests/test_units_mapping.py internally writes to an asdf file in a temporary directory and reads back from there.  Is that enough, or should I add a more explicit test?

Reply:
The latest commit makes `None` the default when units are missing from the asdf file, so we can serialize the example now:

```python
In [3]: from astropy import units as u
   ...: from astropy.modeling.models import Polynomial1D
   ...: model = Polynomial1D(1, c0=1, c1=2).coerce_units({"x": u.m}, {"y": u.s})
   ...: model(u.Quantity(10, u.m))
Out[3]: <Quantity 21. s>

In [4]: import asdf
   ...: asdf.AsdfFile({"model": model}).write_to("test.asdf")
   ...: deserialized_model = asdf.open("test.asdf")["model"]
   ...: deserialized_model(u.Quantity(10, u.m))
Out[4]: <Quantity 21. s>
```